### PR TITLE
interpolation for sentinel 2 correction parameters

### PR DESCRIPTION
**Pull request by panakouris**

### CODE REVIEW

File: atmcorr/processing.py [hunk 1]
```diff
@@ -43,6 +44,62 @@ def _earthdata_credentials_from_env():
...
+        yield i
+
+
+def _get_tile_index_iter(tiling_shape):
+    return itertools.product(range(tiling_shape[0]), range(tiling_shape[1]))
+
+
+def _interpolate_correction_params(correction_params, geometry_dict):
+    # interpolate xa and xc
+    for field in ['xa', 'xc']:
```
Comment (j08lue) [inline atmcorr/processing.py:68]: This nested for loop could be something to paralellize easily, if there is a significant speed gain to expect from that.

File: atmcorr/processing.py [hunk 2]
```diff
@@ -43,6 +44,62 @@ def _earthdata_credentials_from_env():
...
+        for band in range(correction_params[field].shape[0]):
+            selected = correction_params[field][band, :, :].copy()
```
Comment (j08lue) [inline atmcorr/processing.py:70]: Why the `.copy()`? The data is not modified anywhere, is it?

File: atmcorr/processing.py [hunk 3]
```diff
@@ -43,6 +44,62 @@ def _earthdata_credentials_from_env():
...
+            xyi = np.column_stack([x_idx.ravel(), y_idx.ravel()])
+            subset_mask = np.where(~np.isnan(selected))
+            known_data = []
+            xy = np.zeros((len(subset_mask[0]), 2))
+            for idx,  index in enumerate(zip(*subset_mask)):
+                xy[idx] = index
+                known_data.append(selected[index])
+            interpolated = scipy.interpolate.griddata(xy, known_data, xyi, method='nearest')
+            interpolated = interpolated.reshape((selected.shape))
+            correction_params[field][band, :, :] = interpolated.copy()
```
Comment (j08lue) [inline atmcorr/processing.py:81]: Same here - why `.copy()`?

File: atmcorr/processing.py [hunk 4]
```diff
@@ -43,6 +44,62 @@ def _earthdata_credentials_from_env():
...
+
+    # interploate xb correction params in the segments defined by the sensor azimuth
+    seg_a = geometry_dict['sensor_azimuth'] <= np.mean(geometry_dict['sensor_azimuth'])
+    seg_b = geometry_dict['sensor_azimuth'] > np.mean(geometry_dict['sensor_azimuth'])
```
Comment (j08lue) [inline atmcorr/processing.py:85]:
```suggestion
    mean_azimuth = np.mean(geometry_dict['sensor_azimuth'])
    seg_a = geometry_dict['sensor_azimuth'] <= mean_azimuth
    seg_b = geometry_dict['sensor_azimuth'] > mean_azimuth
```

File: atmcorr/processing.py [hunk 5]
```diff
@@ -43,6 +44,62 @@ def _earthdata_credentials_from_env():
...
+    segments = [seg_a, seg_b]
+    for segment in segments:
+        for band in range(correction_params['xb'].shape[0]):
+            selected = correction_params['xb'][band, :, :].copy()
+            segment_mask = np.where(segment)
+            xyi = np.asarray(list(map(list, zip(*segment_mask))))
+            subset_mask = np.where(~np.isnan(selected) & segment)
```
Comment (j08lue) [inline atmcorr/processing.py:92]: I think we could speed this up with boolean indices and get rid of the for loop and appending data to the `known_data` list. Is it worth (performance-wise) to look into that?

**Review by j08lue — CHANGES_REQUESTED** [review]
Nice implementation of the branch between interpolation and no interpolation and clean code.

There are a bit many `for` loops for my taste. But this code is perhaps not a big performance factor?

File: atmcorr/processing.py (hunk 1 above)
Comment (panakouris) [reply atmcorr/processing.py:68]: I wont spend more time to paralellize the code here, as we discussed the main bottleneck of the code is the calls to the API. The interpolation part is relatively fast already

**Pull request merged**